contracts/src/v0.8/dev/automation/2_1/AutomationRegistrar2_1.sol

```diff
@@ -0,0 +1,518 @@
+ *  keeper registry and then emits approved event to finish the flow automatically without manual intervention.
+ * The idea is to have same interface(functions,events) for UI or anyone using this contract irrespective of auto approve being enabled or not.
+ * they can just listen to `RegistrationRequested` & `RegistrationApproved` events and know the status on registrations.
+ */
+contract AutomationRegistrar2_1 is TypeAndVersionInterface, ConfirmedOwner, ERC677ReceiverInterface {
+  /**
+   * DISABLED: No auto approvals, all new upkeeps should be approved manually.
+   * ENABLED_SENDER_ALLOWLIST: Auto approvals for allowed senders subject to max allowed. Manual for rest.
+   * ENABLED_ALL: Auto approvals for all new upkeeps subject to max allowed.
+   */
+  enum AutoApproveType {
+    DISABLED,
+    ENABLED_SENDER_ALLOWLIST,
+    ENABLED_ALL
+  }
+
+  bytes4 private constant REGISTER_REQUEST_SELECTOR = this.register.selector;
+
+  mapping(bytes32 => PendingRequest) private s_pendingRequests;
+  mapping(uint8 => UpkeepRegistrationState) private s_upkeepTypeConfigs;
+
+  LinkTokenInterface public immutable LINK;
+
+  /**
+   * @notice versions:
+   * - KeeperRegistrar 2.1.0: Update for compatability with registry 2.1.0
+   *                          Add auto approval levels by type
+   * - KeeperRegistrar 2.0.0: Remove source from register
+   *                          Breaks our example of "Register an Upkeep using your own deployed contract"
+   * - KeeperRegistrar 1.1.0: Add functionality for sender allowlist in auto approve
+   *                        : Remove rate limit and add max allowed for auto approve
+   * - KeeperRegistrar 1.0.0: initial release
+   */
+  string public constant override typeAndVersion = "AutomationRegistrar 2.1.0";
+
+  /**
+   * @notice UpkeepRegistrationState stores the auto-approval levels for upkeeps by type
+   * @member autoApproveType the auto approval setting (see enum)
+   * @member autoApproveMaxAllowed the max number of upkeeps that can be auto approved of this type
+   * @member approvedCount the count of upkeeps auto approved of this type
+   */
+  struct UpkeepRegistrationState {
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
+    uint32 approvedCount;
+  }
+
+  /**
+   * @notice UpkeepTypeConfig configures the auto-approval levels for upkeeps by type
+   * @member autoApproveType the auto approval setting (see enum)
+   * @member autoApproveMaxAllowed the max number of upkeeps that can be auto approved of this type
+   */
+  struct UpkeepTypeConfig {
+    uint8 upkeepType;
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
+  }
+
+  struct RegistrarConfig {
+    IKeeperRegistryMaster keeperRegistry;
+    uint96 minLINKJuels;
+  }
+
+  struct PendingRequest {
+    address admin;
+    uint96 balance;
+  }
+
+  struct RegistrationParams {
```

Comment:
we had some stack too deep issues in programatic upkeep creation contracts before. Don't fully remember if it was due to a big struct but worth checking that once and break this down if needed?

Reply:
the stack too deep error is known at compile time, yeah? I'm not sure how it would be an issue if the contract compiles successfulls.

Reply:
it's for a contract which programatically registers upkeeps, i.e. calls registerUpkeep with these params. maybe would be good to add a simple test contract in `test/` folder which will ensure it gets compiled?

Reply:
It's definitely _possible_ to do programmatic registrations, but the margin for extra work might be a little limited. Before writing a test, it would be nice to get clear on what are trying to promise to users. Ex we can define this contract...

```solidity
contract MyContract {
  AutomationRegistrar registrar;

  function myRegisterUpkeep(RegistrationParams calldata requestParams) {
    registrar.registerUpkeep(requestParams);
  }
}
```
... which compiles no problem. But if the user tries to do a lot of additional work inside of `myRegisterUpkeep` then they may encounter the "stack too deep" error. I don't think we want to get into the business of promising a certain amount of wiggle room. I also think a creative dev can always find a way around the error by splitting up their work amongst multiple functions.

Reply:
I agree, if the above code snippet compiles that should be enough imo. I just wasn't sure the simple example would compile given issues we have run into before and that we added new fields